# HG changeset patch
# User Jacob Dawid <jacob.dawid@googlemail.com>
# Date 1311717363 -7200
#      Tue Jul 26 23:56:03 2011 +0200
# Node ID ffc2e1d1ad5fd21988ad8cdfe880bb6d0c5bebdd
# Parent  81eef7babeb5a1ed1ffa27479b742bb4b5e24461
History can be filtered with an input search box while typing.

diff --git a/gui/src/HistoryDockWidget.cpp b/gui/src/HistoryDockWidget.cpp
--- a/gui/src/HistoryDockWidget.cpp
+++ b/gui/src/HistoryDockWidget.cpp
@@ -12,39 +12,52 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
 #include "HistoryDockWidget.h"
-#include <QHBoxLayout>
+#include <QVBoxLayout>
 
 HistoryDockWidget::HistoryDockWidget (QWidget * parent):QDockWidget (parent)
 {
   setObjectName ("HistoryDockWidget");
   construct ();
 }
 
 void
 HistoryDockWidget::construct ()
 {
+  m_sortFilterProxyModel.setSourceModel(OctaveLink::instance ()->historyModel());
   m_historyListView = new QListView (this);
-  m_historyListView->setModel (OctaveLink::instance ()->historyModel());
+  m_historyListView->setModel (&m_sortFilterProxyModel);
   m_historyListView->setAlternatingRowColors (true);
   m_historyListView->setEditTriggers (QAbstractItemView::NoEditTriggers);
-  QHBoxLayout *layout = new QHBoxLayout ();
+  m_filterLineEdit = new QLineEdit (this);
+  QVBoxLayout *layout = new QVBoxLayout ();
 
   setWindowTitle (tr ("Command History"));
   setWidget (new QWidget ());
 
   layout->addWidget (m_historyListView);
+  layout->addWidget (m_filterLineEdit);
   layout->setMargin (2);
 
   widget ()->setLayout (layout);
+
+  connect (m_filterLineEdit, SIGNAL (textEdited (QString)), this, SLOT (setFilter (QString)));
 }
 
 void
 HistoryDockWidget::noticeSettings ()
 {
 
 }
+
+void
+HistoryDockWidget::setFilter(QString filter)
+{
+  m_historyListView->setEnabled (false);
+  m_sortFilterProxyModel.setFilterWildcard ( QString ("*%1*").arg (filter));
+  m_historyListView->setEnabled (true);
+}
diff --git a/gui/src/HistoryDockWidget.h b/gui/src/HistoryDockWidget.h
--- a/gui/src/HistoryDockWidget.h
+++ b/gui/src/HistoryDockWidget.h
@@ -15,31 +15,36 @@
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
 #ifndef HISTORYDOCKWIDGET_H
 #define HISTORYDOCKWIDGET_H
 
 #include <QDockWidget>
+#include <QLineEdit>
 #include <QListView>
+#include <QSortFilterProxyModel>
 #include "OctaveLink.h"
+#include <QThread>
 
 class HistoryDockWidget:public QDockWidget
 {
 Q_OBJECT
 public:
   HistoryDockWidget (QWidget *parent = 0);
   void updateHistory (QStringList history);
 
 public slots:
   /** Tells the widget to notice settings that are probably new. */
   void noticeSettings ();
-
+  void setFilter (QString filter);
 signals:
   void information (QString message);
 
 private:
   void construct ();
   QListView *m_historyListView;
+  QLineEdit *m_filterLineEdit;
+  QSortFilterProxyModel m_sortFilterProxyModel;
 };
 
 #endif // HISTORYDOCKWIDGET_H
